ACPT-1428: Create Import JSON API documentation (#193)
AdobeDocs/commerce-webapi: src/pages/rest/modules/import/index.md

---
title: Import API
description: Import entities into Adobe Commerce using REST.
---

# Import data

The `POST /rest/<store_view_code>/V1/import/csv` and `POST /rest/<store_view_code>/V1/import/json` endpoints provide a RESTful way to import data into Adobe Commerce. They mirror the import capabilities found in the Admin at **System** > Data Transfer > **Import**. These endpoints support the import of following entities:

*  `advanced_pricing`
*  `catalog_product`
*  `customer`
*  `customer_address`
*  `customer_composite`
*  `customer_finance`

> *  `stock_sources`

Author: Yes, both import JSON and import CSV support import of Stock Sources. 
But import of Stock Sources is [Inventory](https://github.com/magento-commerce/inventory) functionality.

### JSON Import

The Import JSON API endpoint is only available in `2.4.7-beta`.

The `import/json` endpoint is designed for JSON data:

*  Convert your CSV into JSON using any trustworthy online converter.
*  When converting CSV to JSON using standard tools or libraries, special characters within the data are typically escaped automatically. Ensure that any manual edits or custom conversion processes handle this escaping appropriately.

### Validation Strategy

A validation strategy is mandatory. Depending on your chosen strategy, the API will either proceed with the import or abort it upon encountering invalid rows.

*  `validation-stop-on-errors`: The API halts the import upon finding the first error. It will return an error message describing the issue and its row number.
*  `validation-skip-errors`: The API imports valid rows unless the total number of errors surpasses the `allowedErrorCount`. If errors go beyond this count, the import is halted (even valid rows will not be processed). The response will contain details of the first error found and its row number.

The `allowedErrorCount` field specifies the maximum allowable error count before terminating the import process.

> ## Import CSV API

Author: I do not think that it's relevant for import CSV API as this API was introduced in 2.4.6 version. 
Only import JSON API is introduced in 2.4.7-beta

        "Sprite Foam Yoga Brick": 1,
        "Sprite Yoga Strap 6 foot": 1
    }
}
```

This structure can be found nested within individual product items in the full payload.

**Payload 6 (Gift card product):**

For a gift card product:

*  The `giftcard_amount` attribute accepts an array of amounts that can be used to purchase the gift card.

```json
{
    "giftcard_open_amount_min": 5,
    "giftcard_open_amount_max": 500,
    "giftcard_amount": [
        25,
        50,
        75,
        100
    ]
}
```

This structure can be found nested within individual product items in the full payload.

**Payload 7 (Multiple select attributes):**

Product multiple select attribute is represented as an array of strings.

```json
{
    "additional_attributes": {
        "usage": [
            "Travel",
            "Sports",
            "Office",
            "Casual"
        ]
    }
}
```

Customer and customer address multiple select attributes are represented as an array of strings.

```json
{
    "interests": [
        "Electronics",
        "Sports",
        "Automotive"
    ],
    "delivery_preferences": [
        "Hand directly to resident",
        "Leave in the mailbox"
    ]
}
```

> Import JSON API does not create attributes automatically. You need to create attributes manually before importing data. For more information, see [Create attributes](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/attributes.html#create-attributes).

Reviewer: ```suggestion
The Import JSON API does not create attributes automatically. You need to create attributes manually before importing data. For more information, see [Create attributes](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/attributes.html#create-attributes).
```

```json
{
    "source": {
        "entity": "advanced_pricing",
        "behavior": "append",
        "validation_strategy": "validation-stop-on-errors",
        "allowed_error_count": 10,
        "items":	[
            {
                "sku": "Simple Product 1",
                "tier_price_website": "All Websites [USD]",
                "tier_price_customer_group": "ALL GROUPS",
                "tier_price_qty": 1,
                "tier_price": 7.99,
                "tier_price_value_type": "Fixed"
            },
            {
                "sku": "Simple Product 1",
                "tier_price_website": "base",
                "tier_price_customer_group": "NOT LOGGED IN",
                "tier_price_qty": 2,
                "tier_price": 40,
                "tier_price_value_type": "Discount"
            },
            {
                "sku": "Simple Product 1",
                "tier_price_website": "french",
                "tier_price_customer_group": "General",
                "tier_price_qty": 5,
                "tier_price": 60,
                "tier_price_value_type": "Discount"
            }
        ]
    }
}
```

**Payload (Customer Finances):**

> Customer finances is represented as an array of JSON objects.

Reviewer: ```suggestion
`customer_finances` is represented as an array of JSON objects.
```

Author: If you propose to use entity name here, it's correct to use `customer_finance` instead of `customer_finances`. Fixed and committed.

                "suffix": "",
                "taxvat": "",
                "website_id": 1,
                "password": "",
                "_address_city": "Los Angeles",
                "_address_company": "",
                "_address_country_id": "US",
                "_address_fax": "",
                "_address_firstname": "John",
                "_address_lastname": "Doe",
                "_address_middlename": "",
                "_address_postcode": 90017,
                "_address_prefix": "",
                "_address_region": "California",
                "_address_street": "Main Street 100\nSuite 200",
                "_address_suffix": "",
                "_address_telephone": 3233361267,
                "_address_vat_id": "",
                "_address_default_billing_": 1,
                "_address_default_shipping_": 1,
                "delivery_preferences": [
                    "Hand directly to resident",
                    "Leave in the mailbox"
                ],
                "interests": [
                    "Electronics",
                    "Sports",
                    "Automotive"
                ]
            }
        ]
    }
}
```

> `delivery_preferences` and `interests` are address and customer multiple select attributes. They are represented as an array of strings.

Reviewer: ```suggestion
`delivery_preferences` and `interests` are multiple select attributes for addresses and customers. They are represented as an array of strings.
```

**Response:**

When Import is successful, the API response will contain the number of entities that were successfully imported.

```json
[
  "Entities Processed: 2"
] 
```

> The `"entity": "customer_composite"` represents a composite entity that includes both a customer and an associated address. Thus, even though the payload contains one item in the `items` array, it results in two entities (customer and address) being processed. Therefore, the response indicates "Entities Processed: 2".

Reviewer: ```suggestion
The `"entity": "customer_composite"` represents a composite entity that includes both a customer and an associated address. While the payload contains one item in the `items` array, it results in two entities (customer and address) being processed. Therefore, the response indicates "Entities Processed: 2".
```

<InlineAlert variant="info" slots="text"/>

The Import JSON API is exclusively available via REST and does not support SOAP. This is because the payload consists of complex JSON objects with nested arrays, which are inherently challenging to represent with the XML structure that SOAP relies upon.

### Performance

> The Import JSON API is designed for high performance and scalability, tailored to handle large-scale data imports. In our testing environment, the API demonstrated the capability to efficiently process up to 100,000 products per minute in parallel mode using multiple threads.

Reviewer: ```suggestion
The Import JSON API is designed for high performance and scalability and tailored to handle large-scale data imports. In our testing environment, the API demonstrated the capability to efficiently process up to 100,000 products per minute in parallel mode, using multiple threads.
```
